File: cpp/structure/graph/GraphOfAdjacencyList.cpp
#include <iostream>
#include <vector>
using namespace std;

class GraphOfAdjacencyList {
public:
    GraphOfAdjacencyList() : graphs(vector<VertexNode *>(64, nullptr)), length(0), capacity(64) {}

    bool isEmpty() {
        return !length;
    }

    void addEdge(int index, int edgeIndex, int weight) {
        if (index < 0 || index >= length || edgeIndex < 0 || edgeIndex >= length || index == edgeIndex) {
            return;
        }

        if (!graphs[index]->next) {
            graphs[index]->next = new EdgeNode(edgeIndex, weight);
            return;
        }

        EdgeNode *edge = graphs[index]->next;
        while (edge->next) {
            if (edgeIndex == edge->index) {
                edge->weight = weight;
                return;
            }
            edge = edge->next;
        }
        if (edgeIndex == edge->index) {
            edge->weight = weight;
        } else {
            edge->next = new EdgeNode(edgeIndex, weight);
        }
    }

    void addVertex(int value) {
        if (length >= capacity) {
            graphs.resize(capacity + 16, nullptr);
            capacity += 16;
        }

        graphs[length] = new VertexNode(value);
        ++length;
    }

    void eraseEdge(int index, int edgeIndex) {
        if (index < 0 || index >= length || edgeIndex < 0 || edgeIndex >= length || index == edgeIndex) {
            return;
        }

        EdgeNode *edge = graphs[index]->next;
        if (!edge) {
            return;
        }
        if (edgeIndex == edge->index) {
            graphs[index]->next = edge->next;
            return;
        }

        while (edge->next) {
            if (edgeIndex == edge->next->index) {
                edge->next = edge->next->next;
                return;
            }
            edge = edge->next;
        }
    }

    void eraseVertex(int index) {
        if (index < 0 || index >= length) {
            return;
        }

        for (int i = 0; i < length; ++i) {
            if (i != index) {
                eraseEdge(i, index);
            }
        }

        for (int i = index + 1; i < length; ++i) {
            graphs[i - 1] = graphs[i];
        }
        --length;
        graphs[length] = nullptr;
    }

    void setEdge(int index, int edgeIndex, int weight) {
        if (index < 0 || index >= length || edgeIndex < 0 || edgeIndex >= length || index == edgeIndex) {
            return;
        }

        EdgeNode *edge = graphs[index]->next;
        while (edge && edgeIndex != edge->index) {
            edge = edge->next;
        }

        if (edge) {
            edge->weight = weight;
        }
    }

    void setVertex(int index, int value) {
        if (index >= 0 && index < length) {
            graphs[index]->value = value;
        }
    }

    int getEdge(int index, int edgeIndex) {
        if (index < 0 || index >= length || edgeIndex < 0 || edgeIndex >= length || index == edgeIndex) {
            return INT_MIN;
        }

        EdgeNode *edge = graphs[index]->next;
        while (edge && edgeIndex != edge->index) {
            edge = edge->next;
        }

        if (edge) {
            return edge->weight;
        }
        return INT_MIN;
    }

    int getVertex(int index) {
        if (index >= 0 && index < length) {
            return graphs[index]->value;
        }
        return INT_MIN;
    }

    void clear() {
        graphs.assign(capacity, nullptr);
        length = 0;
    }

private:
    struct EdgeNode {
        int index;
        int weight;
        EdgeNode *next;

        EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
    };

    struct VertexNode {
        int value;
        EdgeNode *next;

        VertexNode(int value) : value(value), next(nullptr) {}
    };

    vector<VertexNode *> graphs;
    int length;
    int capacity;
};

File: py/structure/graph/GraphOfAdjacencyList.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class GraphOfAdjacencyList:
    def __init__(self):
        self.__graphs, self.__length, self.__capacity = [None for _ in range(64)], 0, 64

    def isEmpty(self) -> bool:
        return not self.__length

    def addEdge(self, index: int, edgeIndex: int, weight: int) -> None:
        if index < 0 or index >= self.__length or edgeIndex < 0 or edgeIndex >= self.__length or index == edgeIndex:
            return

        if not self.__graphs[index].next:
            self.__graphs[index].next = self.__EdgeNode(edgeIndex, weight)
            return

        edge = self.__graphs[index].next
        while edge.next:
            if edgeIndex == edge.index:
                edge.weight = weight
                return
            edge = edge.next
        if edgeIndex == edge.index:
            edge.weight = weight
        else:
            edge.next = self.__EdgeNode(edgeIndex, weight)

    def addVertex(self, value: int) -> None:
        if self.__length >= self.__capacity:
            self.__graphs.extend([None for _ in range(16)])
            self.__capacity += 16

        self.__graphs[self.__length] = self.__VertexNode(value)
        self.__length += 1

    def eraseEdge(self, index: int, edgeIndex: int) -> None:
        if index < 0 or index >= self.__length or edgeIndex < 0 or edgeIndex >= self.__length or index == edgeIndex:
            return

        edge = self.__graphs[index].next
        if not edge:
            return
        if edgeIndex == edge.index:
            self.__graphs[index].next = edge.next
            return

        while edge.next:
            if edgeIndex == edge.next.index:
                edge.next = edge.next.next
                return
            edge = edge.next

    def eraseVertex(self, index: int) -> None:
        if index < 0 or index >= self.__length:
            return

        for i in range(len(self.__graphs)):
            if i != index:
                self.eraseEdge(i, index)

        for i in range(index + 1, self.__length):
            self.__graphs[i - 1] = self.__graphs[i]
        self.__length -= 1
        self.__graphs[self.__length] = None

    def setEdge(self, index: int, edgeIndex: int, weight: int) -> None:
        if index < 0 or index >= self.__length or edgeIndex < 0 or edgeIndex >= self.__length or index == edgeIndex:
            return

        edge = self.__graphs[index].next
        while edge and edgeIndex != edge.index:
            edge = edge.next

        if edge:
            edge.weight = weight

    def setVertex(self, index: int, value: int) -> None:
        if index >= 0 and index < self.__length:
            self.__graphs[index].value = value

    def getEdge(self, index: int, edgeIndex: int) -> int:
        if index < 0 or index >= self.__length or edgeIndex < 0 or edgeIndex >= self.__length or index == edgeIndex:
            return -sys.maxsize

        edge = self.__graphs[index].next
        while edge and edgeIndex != edge.index:
            edge = edge.next

        if edge:
            return edge.weight
        return -sys.maxsize

    def getVertex(self, index: int) -> int:
        if index >= 0 and index < self.__length:
            return self.__graphs[index].value
        return -sys.maxsize

    def clear(self) -> None:
        self.__values, self.__length = [None for _ in range(self.__capacity)], 0

    class __EdgeNode:
        def __init__(self, index: int, weight: int):
            self.index, self.weight, self.next = index, weight, None

    class __VertexNode:
        def __init__(self, value: int):
            self.value, self.next = value, None
